mixer/adapter/stdio/zap.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stdio // import "istio.io/istio/mixer/adapter/stdio"
+
+import (
+	"time"
+
+	"github.com/natefinch/lumberjack"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"os"
+
+	"istio.io/istio/mixer/adapter/stdio/config"
+)
+
+var levelToZap = map[config.Params_Level]zapcore.Level{
+	config.INFO:    zapcore.InfoLevel,
+	config.WARNING: zapcore.WarnLevel,
+	config.ERROR:   zapcore.ErrorLevel,
+}
+
+// Creates a zap logger based on the given options
+func newZapLogger(options *config.Params) (*zap.Logger, func(), error) {
+	encCfg := zapcore.EncoderConfig{
+		TimeKey:        "time",
+		LevelKey:       "level",
+		NameKey:        "instance",
+		CallerKey:      "caller",
+		StacktraceKey:  "stack",
+		LineEnding:     zapcore.DefaultLineEnding,
+		EncodeLevel:    zapcore.LowercaseLevelEncoder,
+		EncodeCaller:   zapcore.ShortCallerEncoder,
+		EncodeDuration: zapcore.StringDurationEncoder,
+		EncodeTime:     formatDate,
+	}
+
+	var enc zapcore.Encoder
+	if options.OutputAsJson {
+		enc = zapcore.NewJSONEncoder(encCfg)
+	} else {
+		enc = zapcore.NewConsoleEncoder(encCfg)
+	}
+
+	var outputSink zapcore.WriteSyncer
+	var err error
+	var closer = func() {}
+
+	switch options.LogStream {
+	case config.ROTATED_FILE:
+		lj := &lumberjack.Logger{
+			Filename:   options.OutputPath,
+			MaxSize:    int(options.MaxMegabytesBeforeRotation),
+			MaxBackups: int(options.MaxDaysBeforeRotation),
+			MaxAge:     int(options.MaxRotatedFiles),
+		}
+
+		outputSink = zapcore.AddSync(lj)
+		closer = func() { _ = lj.Close() }
+
+	case config.FILE:
+		if outputSink, closer, err = zap.Open(options.OutputPath); err != nil {
+			return nil, nil, err
+		}
+
+	case config.STDOUT:
+		outputSink = os.Stdout
+
+	case config.STDERR:
+		outputSink = os.Stderr
+	}
+
+	l := zap.New(
+		zapcore.NewCore(enc, outputSink, zap.NewAtomicLevelAt(levelToZap[options.OutputLevel])),
+		zap.ErrorOutput(os.Stderr))
+
+	return l, closer, nil
+}
+
+func formatDate(t time.Time, enc zapcore.PrimitiveArrayEncoder) {
```

Comment:
this looks familiar. possibility for reuse ?

Reply:
Yes, this was stolen from pkg/log. Do you think it's worth exporting the function from pkg/log? It doesn't really belong in the public surface of pkg/log though. And I don't think it warrants a whole new date formatting package...